projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,294 @@
+        if len(datagram_bytearray) != MIN_BYTEARRAY_SIZE + num_node_ids + data_size:
+            raise DatagramTypeError("Invalid Datagram format from bytearray: Not enough data bytes")
+
+        data = datagram_bytearray[NODE_ID_OFFSET + num_node_ids + DATA_SIZE_SIZE:]
+
+        exp_crc32 = cls._calculate_crc32(cls, datagram_type_id, node_ids, data)
+
+        if not exp_crc32 == crc32:
+            raise DatagramTypeError("Invalid Datagram format from bytearray: Invalid crc32")
+
+        return cls(protocol_version=protocol_version,
+                   datagram_type_id=datagram_type_id, node_ids=node_ids, data=data)
+
+    def serialize(self):
+        """This function returns a bytearray based on set data."""
+
+        node_crc32 = zlib.crc32(bytearray(self._node_ids))
+        node_crc32 = self._convert_to_bytearray(node_crc32, 4)
+        data_crc32 = zlib.crc32(self._data)
+        data_crc32 = self._convert_to_bytearray(data_crc32, 4)
+
+        crc32_array = bytearray([self._datagram_type_id,
+                                 len(self._node_ids),
+                                 *node_crc32,
+                                 len(self._data) & 0xff,
+                                 (len(self._data) >> 8) & 0xff,
+                                 *data_crc32])
+        # Update the crc32
+        crc32 = zlib.crc32(crc32_array)
+        crc32 = self._convert_to_bytearray(crc32, 4)
+
+        # Update the bytearray
+        return bytearray([self._protocol_version,
+                          *crc32,
+                          self._datagram_type_id,
+                          len(self._node_ids),
+                          *(self._node_ids),
+                          len(self._data) & 0xff,
+                          (len(self._data) >> 8) & 0xff,
+                          *(self._data)])
+
+    # Accessors and mutators for the datagram
+
+    @property
+    def protocol_version(self):
+        """This function describes the protocol version property."""
+        return self._protocol_version
+
+    @protocol_version.setter
+    def protocol_version(self, value):
+        """This function sets the protocol version."""
+        assert value & 0xff == value
+        self._protocol_version = value & 0xff
+
+    @property
+    def datagram_type_id(self):
+        """This function describes the datagram type id property."""
+        return self._datagram_type_id
+
+    @datagram_type_id.setter
+    def datagram_type_id(self, value):
+        """This function sets the datagram type id."""
+        assert value & 0xff == value
+        self._datagram_type_id = value & 0xff
+
+    @property
+    def node_ids(self):
+        """This function describes the node ids property."""
+        return self._node_ids
+
+    @node_ids.setter
+    def node_ids(self, value):
+        """This function sets the node ids."""
+        assert isinstance(value, list)
+        for val in value:
+            assert val & 0xff == val
+            assert val < 64
```

Comment:
Cool Python thing you can do: `assert all(val & 0xff == val and val < 64 for val in value)`. Also I think `0 <= val < 0xff` would be a nicer way to check those conditions (and Python allows you to chain conditionals like that).